src/QsCompiler/CompilerOptimizations/FunctionEvaluation.fs

```diff
@@ -6,153 +6,147 @@ open Microsoft.Quantum.QsCompiler.DataTypes
-                va |> Seq.map (fun loopValue ->
-                        vars.enterScope()
-                        fillVars vars (StringTuple.fromSymbolTuple (fst s.LoopItem), loopValue.Expression)
-                        let result = this.evaluateScope vars true s.Body
-                        vars.exitScope()
-                        result) |>
-                    Seq.tryFind isInterrupt |? Normal
-            | _ ->
-                "Unknown IterationValue in for loop: " + (printExpr iterValues) |> CouldNotEvaluate
-        | QsWhileStatement s ->
-            Seq.initInfinite (fun _ -> this.evaluateExpression vars s.Condition |> castToBool) |>
-                Seq.map (function
-                    | ExprValue true ->
-                        match this.evaluateScope vars true s.Body with
-                        | Normal -> Normal, false
-                        | res -> res, true
-                    | ExprValue false -> Normal, true
-                    | ExprError s -> CouldNotEvaluate s, true) |>
-                Seq.tryFind snd |? (Normal, true) |> fst
-        | QsRepeatStatement s ->
-            vars.enterScope()
-            let res =
-                Seq.initInfinite (fun _ -> this.evaluateScope vars false s.RepeatBlock.Body) |>
-                Seq.map (function
-                    | Normal ->
-                        match this.evaluateExpression vars s.SuccessCondition |> castToBool with
-                        | ExprValue true -> Normal, true
-                        | ExprValue false ->
-                            match this.evaluateScope vars false s.FixupBlock.Body with
-                            | Normal -> Normal, false
-                            | res -> res, true
-                        | ExprError s -> CouldNotEvaluate s, true
-                    | res -> res, true) |>
-                Seq.tryFind snd |? (Normal, true) |> fst
-            vars.exitScope()
-            res
+            let iterValues = this.evaluateExpression state s.IterationValues
+            let iterSeqOpt =
+                match iterValues with
+                | RangeLiteral _ -> rangeLiteralToSeq iterValues |> Seq.map IntLiteral |> Some
+                | ValueArray va -> va |> Seq.map (fun x -> x.Expression) |> Some
+                | _ -> None
+            match iterSeqOpt with
+            | None -> 
+                "Unknown IterationValue in for loop: " + (printExpr iterValues) |> CouldNotEvaluate |> Error
+            | Some iterSeq ->
+                iterSeq |> Seq.fold (fun result loopValue ->
+                    match result with
+                    | Ok state ->
+                        let state = enterScope state
+                        let state = defineVarTuple state (fst s.LoopItem, loopValue)
+                        let result = this.evaluateScope state true s.Body
+                        match result with
+                        | Ok state -> exitScope state |> Ok
+                        | x -> x
```

Comment:
The assumption here is that x will never wrap a value that contains state? That seems like something that might easily change, and it would thus be better to exit the scope in any case.

Reply:
We only care about the state of the function evaluation while the function is being evaluated normally. As soon as we reach an interrupt, we want to propagate this interrupt back to the top of the evaluation, and the state of the evaluation is unimportant. The interrupt doesn't even store a state.

Reply:
Sure, but code is easily modified. Maybe someone at some point feels like attaching the state to interrupts.